pkg/trillianclient/manager.go

```diff
@@ -0,0 +1,207 @@
+
+// GetTrillianClient returns a Rekor Trillian client wrapper for the given tree ID.
+func (cm *ClientManager) GetTrillianClient(treeID int64) (*TrillianClient, error) {
+	cm.clientMu.RLock()
+	client, ok := cm.trillianClients[treeID]
+	cm.clientMu.RUnlock()
+	if ok {
+		return client, nil
+	}
+
+	conn, err := cm.getConn(treeID)
+	if err != nil {
+		return nil, err
+	}
+
+	cm.clientMu.Lock()
+	defer cm.clientMu.Unlock()
+	// Double-check after acquiring the write lock.
+	if client, ok = cm.trillianClients[treeID]; ok {
+		return client, nil
+	}
+
+	newClient := newTrillianClient(trillian.NewTrillianLogClient(conn), treeID)
+	cm.trillianClients[treeID] = newClient
+	return newClient, nil
+}
+
+func CreateAndInitTree(ctx context.Context, config GRPCConfig, defaultServiceConfig string) (*trillian.Tree, error) {
+	newConn, err := dial(config.Address, config.Port, config.TLSCACert, config.UseSystemTrustStore, defaultServiceConfig)
+	if err != nil {
+		return nil, err
+	}
+	adminClient := trillian.NewTrillianAdminClient(newConn)
+
+	t, err := adminClient.CreateTree(ctx, &trillian.CreateTreeRequest{
+		Tree: &trillian.Tree{
+			TreeType:        trillian.TreeType_LOG,
+			TreeState:       trillian.TreeState_ACTIVE,
+			MaxRootDuration: durationpb.New(time.Hour),
+		},
+	})
+	if err != nil {
+		return nil, fmt.Errorf("create tree: %w", err)
+	}
+	logClient := trillian.NewTrillianLogClient(newConn)
+
+	if err := client.InitLog(ctx, t, logClient); err != nil {
+		return nil, fmt.Errorf("init log: %w", err)
+	}
+	log.Logger.Infof("Created new tree with ID: %v", t.TreeId)
+	return t, nil
+}
+
+func dial(hostname string, port uint16, tlsCACertFile string, useSystemTrustStore bool, serviceConfig string) (*grpc.ClientConn, error) {
+	// Set up and test connection to rpc server
+	var creds credentials.TransportCredentials
+	switch {
+	case useSystemTrustStore:
+		creds = credentials.NewTLS(&tls.Config{
+			ServerName: hostname,
+			MinVersion: tls.VersionTLS12,
+		})
+	case tlsCACertFile != "":
+		tlsCaCert, err := os.ReadFile(filepath.Clean(tlsCACertFile))
+		if err != nil {
+			return nil, fmt.Errorf("failed to load tls_ca_cert: %w", err)
+		}
+		certPool := x509.NewCertPool()
+		if !certPool.AppendCertsFromPEM(tlsCaCert) {
+			return nil, fmt.Errorf("failed to append CA certificate to pool")
+		}
+		creds = credentials.NewTLS(&tls.Config{
+			ServerName: hostname,
+			RootCAs:    certPool,
+			MinVersion: tls.VersionTLS12,
+		})
+	default:
+		creds = insecure.NewCredentials()
+	}
+
+	opts := []grpc.DialOption{grpc.WithTransportCredentials(creds)}
+	if serviceConfig != "" {
+		opts = append(opts, grpc.WithDefaultServiceConfig(serviceConfig))
+	}
+	conn, err := grpc.NewClient(fmt.Sprintf("%s:%d", hostname, port), opts...)
+	if err != nil {
+		return nil, fmt.Errorf("failed to connect to RPC server: %w", err)
+	}
+
+	return conn, nil
+}
+
+// Close stops clients and closes underlying gRPC connections.
+func (cm *ClientManager) Close() error {
+	var err error
+
```

Comment:
Should you lock `clientMu` as well, in case there's any threads trying to access a client?